RandLAPACK/drivers/rl_cqrrpt.hh

```diff
@@ -246,37 +256,70 @@ int CQRRPT<T>::CQRRPT1(
 
     if(this -> timing) {
         qrcp_t_stop = high_resolution_clock::now();
+        resize_t_start = high_resolution_clock::now();
+    }
+
+    T* R_dat = util::upsize(n * n, R);
+    
+    if(this -> timing) {
+        resize_t_stop = high_resolution_clock::now();
+        resize_t_dur  += duration_cast<microseconds>(resize_t_stop - resize_t_start).count();
         rank_reveal_t_start = high_resolution_clock::now();
     }
 
-    // Find rank
-    int k = n;
+    int64_t k = n;
     int i;
-    for(i = 0; i < n; ++i) {
-        if(std::abs(A_hat_dat[i * d + i]) < this->eps) {
-            k = i;
-            break;
+    if(this->naive_rank_estimate)
+    {
+        for(i = 0; i < n; ++i) {
+            if(std::abs(A_hat_dat[i * d + i]) < this->eps) {
+                k = i;
+                break;
+            }
         }
+        this->rank = k;
+    }
+    else {
+        // Oleg's scheme for rank estimation
+        for(i = 0; i < n; ++i) {
+            // copy over an upper-triangular matrix R
+            // from col-maj to row-maj format
+            blas::copy(i + 1, &A_hat_dat[i * d], 1, &R_dat[i], n);
+        }
+
+        // find l2-norm of the full R
+        T norm_R = lapack::lange(Norm::Fro, n, n, R_dat, n);
+
+        T norm_R_sub = lapack::lange(Norm::Fro, 1, n, &R_dat[(n - 1) * n], 1);
+        // Chek if R is full column rank
+        if ((norm_R_sub > 0.000001 * norm_R))
+        {
+            k = n;
```

Comment:
1. There's a typo in ``Chek``.
2. Describe what you're testing here.

Reply:
Done.